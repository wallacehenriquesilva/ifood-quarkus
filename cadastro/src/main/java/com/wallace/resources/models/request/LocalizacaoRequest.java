package com.wallace.resources.models.request;

import org.eclipse.microprofile.openapi.annotations.media.Schema;

import javax.validation.constraints.NotNull;

public class LocalizacaoRequest {

    @NotNull
    @Schema(description = "Latitude do restaurante", required = true)
    private Double latitude;

    @NotNull
    @Schema(description = "Longitude do restaurante", required = true)
    private Double longitude;

    public LocalizacaoRequest() {
    }

    public LocalizacaoRequest(Double latitude, Double longitude) {
        this.latitude = latitude;
        this.longitude = longitude;
    }

    public Double getLatitude() {
        return latitude;
    }

    public void setLatitude(Double latitude) {
        this.latitude = latitude;
    }

    public Double getLongitude() {
        return longitude;
    }

    public void setLongitude(Double longitude) {
        this.longitude = longitude;
    }
}
