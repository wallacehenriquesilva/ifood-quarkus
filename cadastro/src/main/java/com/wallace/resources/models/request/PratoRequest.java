package com.wallace.resources.models.request;

import org.eclipse.microprofile.openapi.annotations.media.Schema;

import javax.validation.constraints.NotEmpty;
import javax.validation.constraints.NotNull;
import javax.validation.constraints.Size;
import java.math.BigDecimal;

public class PratoRequest {

    @NotEmpty
    @Size(min = 3, max = 50)
    @Schema(description = "Nome do prato", required = true)
    private String nome;

    @NotEmpty
    @Size(min = 5, max = 100)
    @Schema(description = "Descrição do prato", required = true)
    private String descricao;

    @NotNull
    @Schema(description = "Valor do prato", required = true)
    private BigDecimal preco;

    public PratoRequest() {
    }

    public PratoRequest(String nome, String descricao, BigDecimal preco) {
        this.nome = nome;
        this.descricao = descricao;
        this.preco = preco;
    }

    public String getNome() {
        return nome;
    }

    public void setNome(String nome) {
        this.nome = nome;
    }

    public String getDescricao() {
        return descricao;
    }

    public void setDescricao(String descricao) {
        this.descricao = descricao;
    }

    public BigDecimal getPreco() {
        return preco;
    }

    public void setPreco(BigDecimal preco) {
        this.preco = preco;
    }
}
