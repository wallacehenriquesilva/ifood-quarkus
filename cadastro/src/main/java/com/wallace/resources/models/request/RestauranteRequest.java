package com.wallace.resources.models.request;

import com.wallace.entities.RestauranteEntity;
import com.wallace.resources.models.request.validation.ValidModel;
import com.wallace.resources.models.request.validation.ValidRequestModel;
import org.eclipse.microprofile.openapi.annotations.media.Schema;

import javax.validation.ConstraintValidatorContext;
import javax.validation.constraints.NotEmpty;
import javax.validation.constraints.NotNull;
import javax.validation.constraints.Pattern;
import javax.validation.constraints.Size;

@ValidModel
public class RestauranteRequest implements ValidRequestModel {

    @NotEmpty
    @Pattern(regexp = "^[0-9]{2}\\.[0-9]{3}\\.[0-9]{3}\\/[0-9]{4}-[0-9]{2}$", message = "CPNJ inválido.")
    //Usamos o @Schema para documentar a o swagger
    @Schema(description = "CNPJ do restaurante", required = true)
    private String cnpj;

    @NotEmpty
    @Size(min = 3, max = 50)
    @Schema(description = "Nome do restaurante", required = true)
    private String nome;

    @NotEmpty
    @Size(min = 3, max = 50)
    @Schema(description = "Nome do proprietário do restaurante", required = true)
    private String proprietario;

    @NotNull
    @Schema(description = "Localização do restaurante", required = true)
    private LocalizacaoRequest localizacao;


    public RestauranteRequest() {
    }

    public RestauranteRequest(String proprietario, String cnpj, String nome, LocalizacaoRequest localizacao) {
        this.proprietario = proprietario;
        this.cnpj = cnpj;
        this.nome = nome;
        this.localizacao = localizacao;
    }

    public String getProprietario() {
        return proprietario;
    }

    public void setProprietario(String proprietario) {
        this.proprietario = proprietario;
    }

    public String getCnpj() {
        return cnpj;
    }

    public void setCnpj(String cnpj) {
        this.cnpj = cnpj;
    }

    public String getNome() {
        return nome;
    }

    public void setNome(String nome) {
        this.nome = nome;
    }

    public LocalizacaoRequest getLocalizacao() {
        return localizacao;
    }

    public void setLocalizacao(LocalizacaoRequest localizacao) {
        this.localizacao = localizacao;
    }


    /*
    Criação do validador, que vai validar assim que recebermos uma requisição, se o CNPJ já esta cadastrado.
     */
    @Override
    public boolean isValid(ConstraintValidatorContext validatorContext) {

        if (RestauranteEntity.find("cnpj", cnpj).count() > 0) {
            validatorContext.disableDefaultConstraintViolation();

            validatorContext.buildConstraintViolationWithTemplate("CNPJ já cadastrado")
                    .addPropertyNode("cnpj")
                    .addConstraintViolation();
            return false;
        }
        return true;
    }
}
