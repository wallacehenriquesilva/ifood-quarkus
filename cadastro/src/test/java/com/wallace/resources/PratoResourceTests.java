package com.wallace.resources;

import com.github.database.rider.cdi.api.DBRider;
import com.github.database.rider.core.api.configuration.DBUnit;
import com.github.database.rider.core.api.configuration.Orthography;
import com.github.database.rider.core.api.dataset.DataSet;
import com.github.database.rider.core.api.dataset.ExpectedDataSet;
import com.wallace.configurations.CadastroLifeCycleManager;
import com.wallace.resources.models.request.PratoRequest;
import io.quarkus.test.common.QuarkusTestResource;
import io.quarkus.test.junit.QuarkusTest;
import io.restassured.RestAssured;
import org.approvaltests.Approvals;
import org.junit.jupiter.api.Test;

import javax.ws.rs.core.Response;
import java.math.BigDecimal;

@DBRider
//Para carregar datasets e inseri-los em nosso banco. Ele é feito em cima do DBUnit, então, podemos definir uma estratégia ao DBUnit
@DBUnit(caseInsensitiveStrategy = Orthography.LOWERCASE) //Dizemos para usar tudo como lowercase
@QuarkusTest
@QuarkusTestResource(CadastroLifeCycleManager.class)
        //Chama o lifecycle para subir o container do testsContainers
class PratoResourceTests {


    @Test
    @DataSet(value = "datasets/pratos-cenario-1.yml", cleanAfter = true)
    /**
     *
     *  @DataSet -> Da para configurarmos varias coisas, como o que sera executado no começo, no fim, etc.
     *  Podemos usar também o @ExpectedDataSet() passando o arquivo do dataset esperado após as ações do teste.
     */
    void getAllPratos() {
        final String result = RestAssured.given()
                .when().get("/api/v1/restaurantes/123/pratos")
                .then()
                .statusCode(Response.Status.OK.getStatusCode())
                .extract().asString();
        //Valida o retorno se ele e exatamente igual ao que temos que salvar em um arquivo com o padrao
        //nome da classe . (ponto) nome do metodo . (ponto) approved.json
        Approvals.verifyJson(result);
    }

    @Test
    @DataSet("datasets/pratos-cenario-1.yml") //Dataset inicial do teste
    @ExpectedDataSet("datasets/responses/pratos-get-one.yml")
        //Valida de acordo com o dataset de retorno
    void getOneRestaurant() {
        final String result = RestAssured.given()
                .when()
                .get("/api/v1/restaurantes/123/pratos/123")
                .then()
                .statusCode(Response.Status.OK.getStatusCode())
                .extract().asString();
    }


    @Test
    @DataSet(value = "datasets/pratos-cenario-1.yml", cleanAfter = true)
    @ExpectedDataSet("datasets/responses/pratos-create.yml")
        //Valida de acordo com o dataset de retorno se criou na base de dados o novo prato
        //A sequencia das informações na lista também é validada
    void createRestaurant() {
        final PratoRequest pratoRequest = new PratoRequest(
                "Prato 2",
                "Segundo prato",
                new BigDecimal("100.11")
        );

        RestAssured.given()
                .when()
                .header("Content-Type", "application/json")
                .body(pratoRequest)
                .post("/api/v1/restaurantes/123/pratos")
                .then()
                .statusCode(Response.Status.CREATED.getStatusCode());
    }


    @Test
    @DataSet(value = "datasets/pratos-cenario-1.yml", cleanAfter = true)
    @ExpectedDataSet("datasets/responses/pratos-update.yml")
        //Valida de acordo com o dataset de retorno se realmente alterou no banco
    void updatePrato() {
        final PratoRequest pratoRequest = new PratoRequest(
                "Prato 1 update",
                "Primeiro prato do restaurante atualizado",
                new BigDecimal("52.00")
        );

        RestAssured.given()
                .when()
                .header("Content-Type", "application/json")
                .body(pratoRequest)
                .put("/api/v1/restaurantes/123/pratos/123")
                .then()
                .statusCode(Response.Status.NO_CONTENT.getStatusCode());
    }

    @Test
    @DataSet(value = "datasets/pratos-cenario-1.yml", cleanAfter = true)
    @ExpectedDataSet("datasets/responses/pratos-delete.yml")
    void deletePrato() {
        RestAssured.given()
                .when()
                .header("Content-Type", "application/json")
                .delete("/api/v1/restaurantes/123/pratos/123")
                .then()
                .statusCode(Response.Status.NO_CONTENT.getStatusCode());

    }

    //TODO -> FAZER TESTES DE ERRO

}
