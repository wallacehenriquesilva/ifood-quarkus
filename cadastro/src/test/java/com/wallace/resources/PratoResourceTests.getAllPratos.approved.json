[
  {
    "descricao": "Primeiro prato do restaurante",
    "id": 123,
    "nome": "Prato 1",
    "preco": 52.98,
    "restaurante": {
      "id": 123,
      "nome": "Restaurante 1",
      "proprietario": "Wallace"
    }
  }
]